### PR TITLE
Limited the capacity of ConsumerWorkService's WorkPools

### PR DESCRIPTION
**Pull request by or150**
A proposed fix for #301 

I've replaced the ConcurrentQueue inside WorkPool with a BlockingCollection in order to be able to limit the amount of work that a WorkPool can contain, and made it configurable.
Every call to AddWork when the pool is full will block.

### CODE REVIEW

File: projects/client/RabbitMQ.Client/src/client/api/ConnectionFactory.cs [hunk 1]
```diff
@@ -142,6 +142,12 @@ public class ConnectionFactory : ConnectionFactoryBase, IConnectionFactory
         public const string DefaultVHost = "/";
 
         /// <summary>
+        /// Default capacity of the work pool buffer, with zero meaning unlimited (value: 100).
+        /// </summary>
+        /// <remarks> PLEASE KEEP THIS MATCHING THE DOC ABOVE.</remarks>
+        public const int DefaultWorkPoolCapacity = 100;
```
Comment (michaelklishin) [inline projects/client/RabbitMQ.Client/src/client/api/ConnectionFactory.cs]: Please rename this to `DefaultConsumerWorkPoolCapacity` for clarity.

**Comment by michaelklishin** [comment]
Thank you for submitting this PR. We will review it when our team is back from a holiday break (around Jan 3rd).

**Comment by bording** [comment]
@or150 The changes proposed here will undo a lot of the work introduced in #253 to remove lock contention and blocking, and I'd hate to see performance benefits it brought be undone.

**Comment by michaelklishin** [comment]
@bording good point. Given that manual acknowledgements and QoS are sufficient to prevent consumer overload, I agree with you.

**Comment by or150** [comment]
After taking another look at it, it would probably cause more harm than good and introduce some new bugs.
I'll try using manual acks instead, it's probably a better way to handle this.

**Pull request closed without merging**